### DISCUSSION TITLE
Code location instability - gRPC Error code: UNAVAILABLE

### DISCUSSION THREAD

**Discussion by ytoast** [dagster-io/dagster#17729] [Troubleshooting FAQs] [answered]
I've recently increased the number of concurrent jobs running on Dagster, and I've noticed that the code location has become less stable than before. I'm encountering errors such as `dagster._core.errors.DagsterUserCodeUnreachableError: Could not reach the user code server. gRPC Error code: UNAVAILABLE`. It seems that the gRPC server connection intermittently disconnects, approximately every 10-20 minutes. This connectivity issue is causing existing jobs to abruptly fail with `dagster._core.errors.DagsterExecutionInterruptedError` errors. It seems that I'm getting this error bcos it is failing the `dagster api grpc-health-check`

What can I do to enhance the stability of the code location?

**Comment by jamiedemaria** [comment, marked as answer]
Hi @ytoast . The `dagster._core.errors.DagsterUserCodeUnreachableError: Could not reach the user code server. gRPC Error code: UNAVAILABLE` means that the process trying to connect to the user code server couldn't. This could be due to any number of reasons depending on your deployment setup. As for the job failures - the executing jobs don't communicate with the user code server, so the `UNAVAILABLE` error isn't causing the job failures, but the cause of the `UNAVAILABLE` error could also be causing the jobs to fail. 

You'll likely need to use the debugging tools for your particular deployment type to debug why the processes are failing

**Reply by mlarose** [reply]
Also, see these comments below to drill down on cause and for possible solutions:

- https://github.com/dagster-io/dagster/discussions/17729#discussioncomment-8855332
- https://github.com/dagster-io/dagster/discussions/17729#discussioncomment-10337923

**Comment by nadamakram** [comment]
hi @jamiedemaria :)

I am facing the same issue on my EKS deployment. 

I am using 2 deployments of user code, one with the default image "docker.io/dagster/dagster-celery-k8s" and one is mine. The default user code is `Running` pod and my one is `CrashLoopBackOff` pod.

The exact error message coming from the Dagit is ```grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "failed to connect to all addresses; last error: UNKNOWN: ipv4:172.20.189.35:3030: Failed to connect to remote host: Connection refused"
	debug_error_string = "UNKNOWN:failed to connect to all addresses; last error: UNKNOWN: ipv4:172.20.189.35:3030: Failed to connect to remote host: Connection refused {grpc_status:14```

the 172.20.189.35 ip is the service ip of my custom user code, I don’t think my issue is related to networking because it connects with the service of the default user code.

I tried to run my image on my local setup by `docker run -it my-image /bin/sh` and then try `dagster dev` and `dagster api grpc --module-name my-module-name --host 0.0.0.0 --port 3030`and it is working properly! and `dagster api grpc-health-check -p 3030` gives gRPC connection successful

so what could be the problem? I can’t get useful info from `kubectl describe`

**Comment by pablo-statsig** [comment]
We have also been running into this periodically, as far as I can tell the code deployments are fine the entire time so it feels weird that the webserver cant seem to connect to it.

**Comment by maxwelljoslyn** [comment]
Checking in representing another team who has been having this problem: irregular, but frequent (several times per week) instances of the GRPC server becoming unreachable, with no other apparent problems. Leading to us having to not only restart Dagster but also go in and manually terminate jobs from the old instance (which Dagster will then "fail" to terminate, saying they're unreachable - but at least then they stop being erroneously blue on the asset graph.)

**Comment by ytoast** [comment]
What kind of debugging tools would you suggest? @jamiedemaria I'm using a k8 deployment setup. I tried snooping around but can't find anything substantial

**Comment by tlarrue** [comment]
@jamiedemaria My team is also experiencing this issue on our k8s deployment. We have around 40 assets actively materializing on 90s interval sensors, which we'd like to scale up to much more. In addition to intermittent gRPC errors, we are also finding the UI to be very slow or unresponsive at times. We are looking for ways to debug or tweak our setup for better stability. We are also trying to understand if there is an upper limit to the number of jobs a k8s Dagster deployment can handle. This is currently a barrier to our team adopting Dagster as a tool.

**Comment by kennycontreras** [comment]
Hello 👋 I'm here also representing my team, which is facing the same issue. 

We have Dagster running on K8s with three pods. One is for the user code, one for dagit, and the last for the daemon/scheduler.

The gRPC error happens randomly, and our pipelines don't fail, but the connection sometimes gets lost when there's a new deployment, and the user code pod needs to be destroyed/created again. To "fix" this, we have to go to the UI and manually reload the workspace. As I said, this doesn't always happen, and considering that we have multiple deployments in a day, the probability is low, but either way, it would be nice to find a solution.

**Comment by mlarose** [comment]
Hello, 

To expand on what was explained by @jamiedemaria, the `DagsterUserCodeUnreachableError` can be emitted by a variety of orchestration routines with the code location, from health checks to launching runs, obtaining assets information, scheduling, sensors, etc.

The underlying causes of this error could include your code location being unhealthy, resource constrained (cpu or memory exhaustion) and unable to answer within a specific timeouts, network issues or underlying infrastructure issues (node cycling, eviction, spot instance termination, etc).

Sometimes one misconfigured node, or malfunctioning essential daemonset on a specific node, can create recurring but not predictable problems at the networking level (dns resolution, setting iptables rules depending on how your Kubernetes is configured to implement ClusterIp services). 

> so what could be the problem? I can’t get useful info from `kubectl describe`

Top of my head:

- `kubectl get events` would give you events for the entire cluster which might yield clues
- `kubectl get pod -o json` could give you information about problems such as container restart, exit code and what node it's running on
- `kubectl logs` to get logs from your container or the `--previous` that might not have been forwarded to dagster
- `kubectl exec` - to get on the pod and examine it from within, or another pod and try to `curl` to the grpc endpoint of your code location to see if it's reachable, or if it can resolve the dns address of the service, which could point to other services in kubernetes such as core-dns being unhealthy.
- `kubectl port-forward` is another way to test if the grpc endpoint is responding


> What kind of debugging tools would you suggest?


- https://k9scli.io/ is fairly user-friendly and perhaps would allow you to diagnose some category of issues.
- some open source or commercial metrics solution to monitor your containers cpu and memory usage, to identify possible problems, like prometheus and grafana
- log collector and aggregation can help understand incidents and problems across services, pods, nodes, etc.
- `py-spy` to perhaps investigate what your code location is doing when/if it's not responsive
- you could preload your code location image with a slew of linux process and network debugging tools   (`dnsutils`, `psutil`, `strace` or `tcpdump`) but for some of these you would also need to add capabilities to your container (like `CAP_SYS_ADMIN`) or run it as a privileged container.


> We are also trying to understand if there is an upper limit to the number of jobs a k8s Dagster deployment can handle. This is currently a barrier to our team adopting Dagster as a tool.

You can hit several soft limit as you scale your usage of Dagster on k8s. How much you compute resources you allocate to the various element of the dagster control plane, kubernetes, etc. How did you configure node scaling for your clusters? How big are your jobs (are they running for hours and consuming tons of cpu or memory) or are they lightweight. Many such soft limits are not Dagster specific but common to running large distributed workloads in general. What are the specific pain points you're encountering?

**Comment by the4thamigo-uk** [comment]
This might be irrelevant, as this error can occur for a lot of reasons, but we have been struggling with this issue on GKE. In our scenario, we occasionally have spot instances that are moved to new GKE nodes. If this causes the dagster postgres instance to be moved, then the new postgres instance on the new node is delayed starting up again because the volume takes up to 5mins to copy. This seems to leave the code location in a broken state and it doesnt seem to recover, even when the postgres finally comes back online. 

Its worth looking at the logs in the daemon process as well as the code location user deployment to see if you are seeing postgres connection errors as an underlying cause.

**Reply by mlarose** [reply]
Thanks for that contribution @the4thamigo-uk. That is a very specific example, but it illustrates a category of infrastructure deployment related problem that might cause this issue indeed and may give insight to others.

**Comment by mlarose** [comment]
A recurring culprit of slow code location server start up is the parsing of dbt models to generate its manifest file. See more about this in: [Loading dbt models from a dbt project](https://docs.dagster.io/integrations/dbt/reference#loading-dbt-models-from-a-dbt-project).

To prevent this, ensure `dbt parse` pregenerates this manifest from CI/CD, either by including this command in your docker image or in [your PEX package if you are using serverless with fast deployment enabled](https://docs.dagster.io/integrations/dbt/using-dbt-with-dagster-plus#step-3-update-the-cicd-files).

**Comment by shayashi-aam** [comment]
Our team has been hit by this issue as well. We are just running "dagster dev" inside a docker container. The error message we see is slightly different from the one that posted by @nadamakram 

```
dagster._core.errors.DagsterUserCodeUnreachableError: Could not reach user code server. gRPC Error code: UNAVAILABLE
  File "/opt/anaconda/2023/envs/pandas2/lib/python3.11/site-packages/dagster/_grpc/client.py", line 657, in start_run
    res = self._query(
          ^^^^^^^^^^^^
  File "/opt/anaconda/2023/envs/pandas2/lib/python3.11/site-packages/dagster/_grpc/client.py", line 216, in _query
    self._raise_grpc_exception(
  File "/opt/anaconda/2023/envs/pandas2/lib/python3.11/site-packages/dagster/_grpc/client.py", line 199, in _raise_grpc_exception
    raise DagsterUserCodeUnreachableError(
The above exception was caused by the following exception:
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "failed to connect to all addresses; last error: UNKNOWN: unix:/tmp/tmpda756w01: connect: No such file or directory (2)"
	debug_error_string = "UNKNOWN:Error received from peer  {created_time:"2024-10-24T00:00:37.058736227+00:00", grpc_status:14, grpc_message:"failed to connect to all addresses; last error: UNKNOWN: unix:/tmp/tmpda756w01: connect: No such file or directory (2)"}"
>
  File "/opt/anaconda/2023/envs/pandas2/lib/python3.11/site-packages/dagster/_grpc/client.py", line 214, in _query
    return self._get_response(method, request=request_type(**kwargs), timeout=timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda/2023/envs/pandas2/lib/python3.11/site-packages/dagster/_grpc/client.py", line 174, in _get_response
    return getattr(stub, method)(request, metadata=self._metadata, timeout=timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda/2023/envs/pandas2/lib/python3.11/site-packages/grpc/_channel.py", line 1181, in __call__
    return _end_unary_response_blocking(state, call, False, None)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/anaconda/2023/envs/pandas2/lib/python3.11/site-packages/grpc/_channel.py", line 1006, in _end_unary_response_blocking
    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I see no spikes in cpu/memory/network usage when this occurs, and the problem disappears after an hour or so by itself. 

We had to set "max_concurrent_runs: 1" to limit the concurrency as each of our asset runs in parallel using ray. I don't think this has anything to do with this issue. Do you recommend setting any special options for docker container (such as shm_size?)

**Comment by kennycontreras** [comment]
Besides the `dbt manifest` suggestion, has anyone been able to find a solution?

I have tried everything, from increasing the local_startup_timeout and reload_time in the `dagster.yaml` like the following:

```yaml
code_servers:
  local_startup_timeout: 300
  reload_timeout: 300
```

to add env variables 

```bash
DAGSTER_GRPC_TIMEOUT_SECONDS=300
DAGSTER_REPOSITORY_GRPC_TIMEOUT_SECONDS=300
```

Still, nothing seems to improve.

**Comment by babaMar** [comment]
For us is now not able to load the code locations and it intermittently fails with:

```
dagster._core.errors.DagsterUserCodeUnreachableError: Could not reach user code server. gRPC Error code: UNAVAILABLE
  File "/usr/local/lib/python3.10/site-packages/dagster/_core/workspace/context.py", line 633, in _load_location
    origin.reload_location(self.instance)
  File "/usr/local/lib/python3.10/site-packages/dagster/_core/remote_representation/origin.py", line 347, in reload_location
    self.create_client().reload_code(timeout=instance.code_server_reload_timeout)
  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/client.py", line 323, in reload_code
    return self._query("ReloadCode", api_pb2.ReloadCodeRequest, timeout=timeout)
  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/client.py", line 173, in _query
    self._raise_grpc_exception(
  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/client.py", line 156, in _raise_grpc_exception
    raise DagsterUserCodeUnreachableError(
The above exception was caused by the following exception:
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "failed to connect to all addresses; last error: UNKNOWN: ipv4:172.20.255.107:3030: Failed to connect to remote host: Connection refused"
	debug_error_string = "UNKNOWN:Error received from peer  {grpc_message:"failed to connect to all addresses; last error: UNKNOWN: ipv4:172.20.255.107:3030: Failed to connect to remote host: Connection refused", grpc_status:14, created_time:"2025-01-29T10:21:29.544143706+00:00"}"
>
  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/client.py", line 171, in _query
    return self._get_response(method, request=request_type(**kwargs), timeout=timeout)
  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/client.py", line 141, in _get_response
    return getattr(stub, method)(request, metadata=self._metadata, timeout=timeout)
  File "/usr/local/lib/python3.10/site-packages/grpc/_channel.py", line 1181, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/usr/local/lib/python3.10/site-packages/grpc/_channel.py", line 1006, in _end_unary_response_blocking
    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable
```

and 

```
dagster._core.errors.DagsterUserCodeUnreachableError: Could not reach user code server. gRPC Error code: UNAVAILABLE

  File "/usr/local/lib/python3.10/site-packages/dagster/_core/workspace/context.py", line 635, in _load_location
    else origin.create_location(self.instance)
  File "/usr/local/lib/python3.10/site-packages/dagster/_core/remote_representation/origin.py", line 363, in create_location
    return GrpcServerCodeLocation(self, instance=instance)
  File "/usr/local/lib/python3.10/site-packages/dagster/_core/remote_representation/code_location.py", line 646, in __init__
    list_repositories_response = sync_list_repositories_grpc(self.client)
  File "/usr/local/lib/python3.10/site-packages/dagster/_api/list_repositories.py", line 20, in sync_list_repositories_grpc
    api_client.list_repositories(),
  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/client.py", line 250, in list_repositories
    res = self._query("ListRepositories", api_pb2.ListRepositoriesRequest)
  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/client.py", line 173, in _query
    self._raise_grpc_exception(
  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/client.py", line 156, in _raise_grpc_exception
    raise DagsterUserCodeUnreachableError(

The above exception was caused by the following exception:
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "failed to connect to all addresses; last error: UNKNOWN: ipv4:172.20.255.107:3030: Failed to connect to remote host: Connection refused"
	debug_error_string = "UNKNOWN:Error received from peer  {created_time:"2025-01-29T10:21:51.526834743+00:00", grpc_status:14, grpc_message:"failed to connect to all addresses; last error: UNKNOWN: ipv4:172.20.255.107:3030: Failed to connect to remote host: Connection refused"}"
>

  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/client.py", line 171, in _query
    return self._get_response(method, request=request_type(**kwargs), timeout=timeout)
  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/client.py", line 141, in _get_response
    return getattr(stub, method)(request, metadata=self._metadata, timeout=timeout)
  File "/usr/local/lib/python3.10/site-packages/grpc/_channel.py", line 1181, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/usr/local/lib/python3.10/site-packages/grpc/_channel.py", line 1006, in _end_unary_response_blocking
    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable

The above exception occurred during handling of the following exception:
dagster._core.errors.DagsterUserCodeUnreachableError: Could not reach user code server. gRPC Error code: UNAVAILABLE

  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/server_watcher.py", line 119, in watch_grpc_server_thread
    watch_for_changes()
  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/server_watcher.py", line 82, in watch_for_changes
    new_server_id = client.get_server_id(timeout=REQUEST_TIMEOUT)
  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/client.py", line 233, in get_server_id
    res = self._query("GetServerId", api_pb2.Empty, timeout=timeout)
  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/client.py", line 173, in _query
    self._raise_grpc_exception(
  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/client.py", line 156, in _raise_grpc_exception
    raise DagsterUserCodeUnreachableError(

The above exception was caused by the following exception:
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "failed to connect to all addresses; last error: UNKNOWN: ipv4:172.20.255.107:3030: Failed to connect to remote host: Connection refused"
	debug_error_string = "UNKNOWN:Error received from peer  {grpc_message:"failed to connect to all addresses; last error: UNKNOWN: ipv4:172.20.255.107:3030: Failed to connect to remote host: Connection refused", grpc_status:14, created_time:"2025-01-29T10:21:40.454812676+00:00"}"
>

  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/client.py", line 171, in _query
    return self._get_response(method, request=request_type(**kwargs), timeout=timeout)
  File "/usr/local/lib/python3.10/site-packages/dagster/_grpc/client.py", line 141, in _get_response
    return getattr(stub, method)(request, metadata=self._metadata, timeout=timeout)
  File "/usr/local/lib/python3.10/site-packages/grpc/_channel.py", line 1181, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/usr/local/lib/python3.10/site-packages/grpc/_channel.py", line 1006, in _end_unary_response_blocking
    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable
```

Looking into Kubernetes events and such, can't really see anything. It seems more related to the pod taking a long time to stabilize.

It seems the process got killed for some reason, these are the last logs from the pod:
```
2025-01-29 09:27:55 +0000 - dagster.code_server - INFO - Stopping server once all current RPC calls terminate or 60 seconds pass
INFO:dagster.code_server:Stopping server once all current RPC calls terminate or 60 seconds pass
2025-01-29 09:28:06 +0000 - dagster.code_server - INFO - Shutting down Dagster code server for file /assets_dbt/assets_dbt/definitions.py in process 1124147
INFO:dagster.code_server:Shutting down Dagster code server for file /assets_dbt/assets_dbt/definitions.py in process 1124147
```

**Reply by KeithHanson** [reply]
This exact issue is happening with our deployment as well. We see the issue crop up in a sensor. Reloading code server seems to fix things but... would love some pointers.

**Reply by KeithHanson** [reply]
Alrighty. After much gnashing of teeth, we found a rogue piece of code that wasn't handling exceptions, leaving connections open, ultimately later after many many errors, unable to open a new connection.

The tip off for us was seeing an error from FreeTDS prior to ever hitting our new try/catch. 

Once we rebooted the entire server (not just the containers), our try/catches are firing properly and disposing of the connections, and we haven't seen the error in the last 30 mins or so (we were seeing it within 3-5 minutes indicating one of our routines was causing the exhaustion).

**Reply by garethbrickman** [reply]
That kind of sensor evaluation error could be due to the code server pod that is running the sensor code becoming unresponsive. Redeploying the code location would likely resolve the problem as a short-term solution by spinning up a brand new pod.

Figuring out why exactly it has stopped serving any requests may require exec'ing into the pod and running `py-spy` dump ([as shown here](https://github.com/dagster-io/dagster/discussions/14771#discussioncomment-6165783)) to see what the threads are doing- a common cause is sporadically hanging code in sensors gradually burning through threads over time.

Note that upgrading the agent may help: as of version 1.7.2 the agent detects this type of situation and automatically redeploys the pod.

**Comment by nicchristeller** [comment]
Some insight, in case it helps anyone here.

I was running containerized, with the "code server" container hosting the dagster API, being launched with `dagster code-server start`*. 

This launches a "proxy" process, and separately launches the "real" GRPC API process. The proxy process passes requests through to the second process.

In my case, a sensor was using too much memory. This caused the "real" process to crash _silently_. 

Unfortunately, the "proxy" process does not detect this, and therefore the webserver/daemon nodes will receive `gRPC Error code: UNAVAILABLE` when attempting to use the code server. Even worse, gRPC healthchecks to the code server continued to pass, because the proxy process was still alive.

Importantly, pressing "Reload" in the UI will cause the "real" process to be recreated, and it will then continue to work until something causes it to crash again.

One improvement here it to use `dagster api grpc` instead of `dagster code-server start`. At least in my case, this means my orchestrator can detect the failed process and report the error (Error code 137 from docker, giving a better hint at what is going on), as well as automatically restart the gRPC server.

**Reply by nicchristeller** [reply]
\* `dagster code-server start` seems to be [undocumented](https://docs.dagster.io/api/python-api/cli). I have copied it from the [deployment guide](https://docs.dagster.io/guides/deploy/deployment-options/docker#define-a-docker-image-for-each-code-location), which I imagine others have too.

**Comment by axellpadilla** [comment]
This continues to be an issue. Do you know if launching the grpc manually helps with this on module instances?, looks like k8s and docker are also having this problem constantly.

**Reply by frankcampos** [reply]
I believe the issues start when we increase the jobs. One of the solutions is to select only the job needed in our local dagster run. We can update the location name in our workspace.yaml.
![image](https://github.com/user-attachments/assets/b4c643e5-cb4a-44e9-b08d-6b62b8f11539)

**Reply by axellpadilla** [reply]
@frankcampos  I need multiple jobs, assets, etc. per location

**Comment by frankcampos** [comment]
I believe the issues start when we increase the jobs. One of the solutions is to select only the job needed in our local dagster run. We can update the location name in our workspace.yaml.
![image](https://github.com/user-attachments/assets/99faa75b-a600-43b1-8286-345fc8b05d4f)

**Comment by axellpadilla** [comment]
Linking apparently related open issues here so devs can prioritize this, maybe, initially adding ways to better debug and know what is actually happening:
https://github.com/dagster-io/dagster/issues/27266
https://github.com/dagster-io/dagster/issues/25116
https://github.com/dagster-io/dagster/issues/24050
https://github.com/dagster-io/dagster/issues/22511
https://github.com/dagster-io/dagster/issues/30007
https://github.com/dagster-io/dagster/issues/18760
https://github.com/dagster-io/dagster/issues/17493
https://github.com/dagster-io/dagster/issues/8073

**Comment by panandr** [comment]
We are also hitting this issue in our dagster deployment in EKS. Has anyone found a workaround/solution for this?

**Reply by babaMar** [reply]
> We are also hitting this issue in our dagster deployment in EKS. Has anyone found a workaround/solution for this?

I ended up using the deployments `affinity` to spread the code locations across nodes and avoiding to deploy on the same node (at the same time setting the resource requests) with pod anti-affinity (or `topologySpreadConstraints` via the same key) on the code-location labels:

```
    affinity:
      podAntiAffinity:
        preferredDuringSchedulingIgnoredDuringExecution:
          - weight: 100                                                                                                
            podAffinityTerm:
              topologyKey: kubernetes.io/hostname
              labelSelector:
                matchLabels:
                  app.kubernetes.io/instance: dagster-user-deployments
```

Not sure if you also have multiple code locations though.

If you're hitting this problem with a single code location, you could try using the startup probe, from the docs: 

"Startup probes are useful for Pods that have containers that take a long time to come into service. Rather than set a long liveness interval, you can configure a separate configuration for probing the container as it starts up, allowing a time longer than the liveness interval would allow.

If your container usually starts in more than **initialDelaySeconds**  + **failureThreshold**  x **periodSeconds**, you should specify a startup probe that checks the same endpoint as the liveness probe. The default for `periodSeconds` is 10s. You should then set its `failureThreshold` high enough to allow the container to start, without changing the default values of the liveness probe."

**Comment by HuseynA28** [comment]
Dagster kept crashing on WSL with "No heartbeat received in 30 seconds." It turned out the problem wasn't Dagster at all — it was WSL. The WSL clock had drifted about 36 seconds behind the Windows clock. Dagster checks time to decide if a process is still alive, so when the clock jumped, it thought the process had gone silent for too long and shut it down. On native Windows this never happens because the time stays correct. Fixing the WSL clock fixed the error.

**Reply by axellpadilla** [reply]
This is actually a pretty useful diagnostic hint that could impact any environment even when just different containers, dagster should better handle the delta to try to always compare datetimes from the same environment (ej the database or daemon) if it doesn't do that
